src/modelfree/common/lookback.py

```diff
@@ -0,0 +1,274 @@
+
+            # update lb_tuple[idx].data since this data will be used elsewhere
+            self.lb_tuples[idx].data['action'] = lb_action
+            self.lb_tuples[idx].data['state'] = lb_state
+            for env_idx in range(self.num_envs):
+                self.lb_tuples[idx].data['info'][env_idx].update(lb_info[env_idx])
+
+    def _reset_state_data(self, initial_observations, env_idx=None):
+        """Reset lb_venv states when self.venv resets. Also reset data for baseline policy.
+
+        :param initial_observations ([float]) observations from freshly reset self.venv
+        :return: None
+        """
+        # get action and state to synchronize caches for lookback agents
+        obs = self._get_truncated_obs(initial_observations)
+        action, state = self._policy.predict(obs, state=None, mask=None, deterministic=True)
+
+        # data from our own environment which needs to be set elsewhere
+        env_data = self.venv.unwrapped.env_method('get_state', indices=env_idx, all_data=True)
+        mj_states, sim_data, radii = list(zip(*env_data))
+
+        for lb_tuple in self.lb_tuples:
+            # synchronize lb_tuple data caches
+            if env_idx is None:
+                # this branch is only called in self.reset()
+                lb_tuple.venv.reset()
+                lb_tuple.data['action'] = action
+                lb_tuple.data['state'] = state
+            else:
+                # this gets called when an episode ends in one of the environments
+                lb_tuple.data['action'][env_idx] = action[env_idx]
+                if state is None:
+                    lb_tuple.data['state'] = None
+                else:
+                    lb_tuple.data['state'][env_idx, :, :] = state[env_idx, :, :]
+
+            # synchronize environment states
+            self._sync_curry_venvs(lb_tuple, env_idx)
+            envs_iter = range(self.num_envs) if env_idx is None else (env_idx,)
+            for i, env_to_set in enumerate(envs_iter):
+                env = lb_tuple.venv.unwrapped
+                env.env_method('set_state', mj_states[i], indices=env_to_set,
+                               sim_data=sim_data[i], radius=radii[i], forward=False)
+
+    def _sync_curry_venvs(self, lb_tuple, env_idx):
+        """Synchronize observation of victim in lookback CurryVecEnv to that of our own victim."""
+        our_curry_obs = self.get_curry_obs(env_idx=env_idx)
+        lb_tuple.venv.set_curry_obs(our_curry_obs, env_idx=env_idx)
+
+    def _get_truncated_obs(self, obs):
+        """Truncate the observation given to self._policy if we are using adversarial noise ball"""
+        if isinstance(self._policy.policy, ActorCriticRLModel):
+            return obs[:, :self._policy.policy.observation_space.shape[0]]  # stable_baselines
+        else:
+            return obs[:, :self._policy.policy.ob_space.shape[0]]           # gym_compete
```

Comment:
Maybe we should just change `gym_compete` to use `observation_space` for consistency?